app/models/atmosphere/appliance.rb

```diff
@@ -129,15 +127,40 @@ def owned_by?(user)
       appliance_set.user_id == user.id
     end
 
+    def prepaid_until
+      # Helper method which determines how long this appliance will remain prepaid
+      if deployments.blank?
+        # Invoking this method on an appl whith 0 deployments is a nasal demons
+        # scenario: the question doesn't make sense and neither will the answer.
+        # Returning fixed time is least likely to break the client.
+        result = Time.parse('2000-01-01 12:00')
+      else
+        result = deployments.max_by { |dep| dep.prepaid_until }.prepaid_until
+      end
+      result
+    end
+
+    # This method provided for backward compatibility
+    def billing_state
+      if deployments.any? { |d| d.billing_state == 'prepaid' }
+        'prepaid'
+      else
+        'expired'
+      end
+    end
+
     private
 
     def assign_default_fund
-      # This is a "goalkeeper" method which will assign this appliance to its owner's default fund, if no fund has been specified yet.
-      # It is provided to ensure compatibility with old APIs of Atmosphere which do not request funds to be specified when instantiating appliances
+      # This is a "goalkeeper" method which will assign this appliance
+      # to its owner's default fund, if no fund has been specified yet.
+      # It is provided to ensure compatibility with old APIs of Atmosphere
+      # which do not request funds to be specified when instantiating appliances
       # Once the APIs have been updated, this method will be deprecated and should be removed.
       if self.fund.blank?
```

Comment:
`self` is redundant for getter methods, it is required only for setters, so it can be changed into:

```ruby
if fund.blank?
  self.fund = appliance_set.user.default_fund
  ...
```